File: Apps/BankBase/backend/src/main/scala/model/session/Session.scala
package model.session

import io.github.zeal18.zio.mongodb.bson.annotations.BsonId
import zio.json.{DeriveJsonCodec, JsonCodec}

import java.time.LocalDateTime
import java.util.UUID

case class Session(
                                   @BsonId _id: String,
                                   openedAt: LocalDateTime,
                                   lastInteractionTime: LocalDateTime,
                                   userAgent: String,
                                   accessToken: String,
                                   refreshToken: String
                                 ) {
  def this(userAgent: String, accessToken: String, refreshToken: String) = {
    this(UUID.randomUUID().toString, LocalDateTime.now(), LocalDateTime.now(), userAgent, accessToken, refreshToken)
  }

  inline given JsonCodec[Session] = DeriveJsonCodec.gen[Session]
}

object Session {
  def apply(userAgent: String, accessToken: String, refreshToken: String): Session =
    new Session(
      userAgent,
      accessToken,
      refreshToken
    )

  inline given JsonCodec[Session] = DeriveJsonCodec.gen
}

case class Sessions(
                     @BsonId _id: String,
                     sessions: List[Session]
                   )

object Sessions {
  inline given JsonCodec[Sessions] = DeriveJsonCodec.gen
}

File: Apps/AuthMS/model/session/session.go
package session

import "time"
import "github.com/google/uuid"

type Session struct {
	id                  string
	openedAt            time.Time
	lastInteractionTime time.Time
	userAgent           string
	accessToken         string
	refreshToken        string
}

func (s *Session) Id() string {
	return s.id
}

func (s *Session) OpenedAt() time.Time {
	return s.openedAt
}

func (s *Session) LastInteractionTime() time.Time {
	return s.lastInteractionTime
}

func (s *Session) UserAgent() string {
	return s.userAgent
}

func (s *Session) AccessToken() string {
	return s.accessToken
}

func (s *Session) RefreshToken() string {
	return s.refreshToken
}

type Sessions struct {
	id       string
	sessions []*Session
}

func (s *Sessions) Id() string {
	return s.id
}

func (s *Sessions) Sessions() []*Session {
	return s.sessions
}

func NewSession(userAgent string, accessToken string, refreshToken string) (*Session, error) {
	sessionId, err := uuid.NewRandom()
	if err != nil {
		return nil, err
	}
	return &Session{
		id:                  sessionId.String(),
		openedAt:            time.Now(),
		lastInteractionTime: time.Now(),
		userAgent:           userAgent,
		accessToken:         accessToken,
		refreshToken:        refreshToken,
	}, nil
}

func SessionConstructor(
	id string,
	openedAt time.Time,
	lastInteractionTime time.Time,
	userAgent string,
	accessToken string,
	refreshToken string,
) *Session {
	return &Session{
		id:                  id,
		openedAt:            openedAt,
		lastInteractionTime: lastInteractionTime,
		userAgent:           userAgent,
		accessToken:         accessToken,
		refreshToken:        refreshToken,
	}
}

func SessionsConstructor(
	id string,
	sessions []*Session,
) *Sessions {
	return &Sessions{
		id:       id,
		sessions: sessions,
	}
}
